packages/time-oracle/src/migrate_v1.rs

```diff
@@ -53,42 +54,35 @@ impl<'a> AlarmsOld<'a> {
         }
     }
 
-    pub fn migrate(&self, storage: &mut dyn Storage) -> Result<(), AlarmError> {
+    pub fn migrate(
+        &self,
+        storage: &mut dyn Storage,
+        new_namespace: &str,
+        new_index: &str,
+    ) -> Result<(), AlarmError> {
         let old_alarms = self.alarms();
 
-        let (alarms, ids) = old_alarms
-            .idx
-            .alarms
-            .range(storage, None, None, Order::Ascending)
-            .try_fold(
-                (vec![], vec![]),
-                |mut v: (Vec<AlarmOld>, Vec<Id>), alarm| -> Result<_, StdError> {
-                    let alarm = alarm?;
-                    v.1.push(alarm.0);
-                    if let Some(last) = v.0.last() {
-                        if last.addr != alarm.1.addr {
-                            v.0.push(alarm.1)
-                        }
-                    } else {
-                        v.0.push(alarm.1)
-                    }
-                    Ok(v)
-                },
-            )?;
-
-        // purge all the data
-        for id in ids {
-            old_alarms.remove(storage, id)?;
-        }
-        self.next_id.remove(storage);
+        let alarms_new = Alarms::new(new_namespace, new_index);
+
+        loop {
+            let Some(result) = old_alarms.idx.alarms.range(
+                storage,
+                None,
+                None,
+                Order::Ascending
+            ).next() else {
+                break;
+            };
 
-        let alarms_new = Alarms::new(self.namespace_alarms, self.namespace_index);
+            let (id, alarm) = result?;
+
+            old_alarms.remove(storage, id)?;
```

Comment:
take a look into `IndexedMap::replace` to save an extra read per alarm. Check out the documentation of that method and take a look at the implementation.